src/coreclr/vm/methodtable.cpp

```diff
@@ -2881,6 +2881,523 @@ void  MethodTable::AssignClassifiedEightByteTypes(SystemVStructRegisterPassingHe
+                    {
+                        size = GetLoongArch64PassStructInRegisterFlags((CORINFO_CLASS_HANDLE)pMethodTable);
+                    }
+                    else if (pMethodTable->GetNumIntroducedInstanceFields() == 2)
+                    {
+                        size = GetLoongArch64PassStructInRegisterFlags((CORINFO_CLASS_HANDLE)pMethodTable);
+                    }
+                }
+                else if (fieldType == ELEMENT_TYPE_CLASS)
+                {
+                    pMethodTable = pFieldStart->GetFieldTypeHandleThrowing().GetMethodTable();
+                    if (pMethodTable->IsArray())
+                        size = GetLoongArch64PassStructInRegisterFlags((CORINFO_CLASS_HANDLE)pMethodTable);
+                }
+            }
+            else if (numIntroducedFields == 2)
+            {
+                FieldDesc *pFieldStart = pMethodTable->GetApproxFieldDescListRaw();
+
+                CorElementType fieldType = pFieldStart[0].GetFieldType();
+                if (CorTypeInfo::IsPrimitiveType_NoThrow(fieldType))
+                {
+                    if (fieldType == ELEMENT_TYPE_R4)
+                        size = STRUCT_FLOAT_FIELD_FIRST;
+                    else if (fieldType == ELEMENT_TYPE_R8)
+                        size = STRUCT_FIRST_FIELD_DOUBLE;
+                    else if (pFieldStart[0].GetSize() == 8)
+                        size = STRUCT_FIRST_FIELD_SIZE_IS8;
+
+                }
+                else if (fieldType == ELEMENT_TYPE_VALUETYPE)
+                {
+                    pMethodTable  = pFieldStart->GetFieldTypeHandleThrowing().GetMethodTable();
+                    if (pMethodTable->GetNumIntroducedInstanceFields() == 1)
+                    {
+                        size = GetLoongArch64PassStructInRegisterFlags((CORINFO_CLASS_HANDLE)pMethodTable);
+                        if (size == STRUCT_FLOAT_FIELD_ONLY_ONE)
+                        {
+                            size = pFieldStart[0].GetSize() == 8 ? STRUCT_FIRST_FIELD_DOUBLE : STRUCT_FLOAT_FIELD_FIRST;
+                        }
+                        else if (size == STRUCT_NO_FLOAT_FIELD)
+                        {
+                            size = pFieldStart[0].GetSize() == 8 ? STRUCT_FIRST_FIELD_SIZE_IS8: 0;
+                        }
+                        else
+                        {
+                            size = STRUCT_NO_FLOAT_FIELD;
+                            goto _End_arg;
+                        }
+                    }
+                    else
+                    {
+                        size = STRUCT_NO_FLOAT_FIELD;
+                        goto _End_arg;
+                    }
+                }
+                else if (fieldType == ELEMENT_TYPE_CLASS)
+                {
+                    MethodTable* pMethodTable2  = pFieldStart->GetFieldTypeHandleThrowing().GetMethodTable();
+                    if (pMethodTable2->IsArray())
```

Comment:
It does not make sense to special case arrays for the field layout. Why did you add special casing arrays here? 

Note that arrays in .NET are regular objects. They are not inline like in C/C++.

Reply:
If always within managed code there is no need to add this, for calling unmanaged code while using LoongArch64's ABI, 
liking
```
[StructLayout(LayoutKind.Sequential)]
public unsafe struct FixedArrayClassificationTest
{
    [MarshalAs(UnmanagedType.ByValArray, SizeConst = 3)] 
    public Int32Wrapper[] arr;
    public float f;
}

[StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
public struct UnicodeCharArrayClassification
{
    [MarshalAs(UnmanagedType.ByValArray, SizeConst = 6)] 
    public char[] arr;
    public float f;
}
```
which the field of struct is array type, LoongArch64's ABI has to parse the field's type.

Reply:
Right, but the code here does not deal with the unmanaged view of the type. It deals with the managed view of the type where the array is regular object reference.

Reply:
After refactor, this interface will be called also by managed from the jitinterface.
It will be called within the file `vm/callingconvention.h`.

Reply:
Right, but that does not explain the special casing of arrays in this spot. Do you have a test case that fails when you delete the special casing of the arrays here?

Reply:
> Right, but that does not explain the special casing of arrays in this spot.


> Do you have a test case that fails when you delete the special casing of the arrays here?

The following code is from the case `baseservices/TieredCompilation/BasicTestWithMcj/BasicTestWithMcj.sh`

> ```
> [StructLayout(LayoutKind.Sequential)]
> public unsafe struct FixedArrayClassificationTest
> {
>     [MarshalAs(UnmanagedType.ByValArray, SizeConst = 3)] 
>     public Int32Wrapper[] arr;
>     public float f;
> }
> 
> [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
> public struct UnicodeCharArrayClassification
> {
>     [MarshalAs(UnmanagedType.ByValArray, SizeConst = 6)] 
>     public char[] arr;
>     public float f;
> }
> ```

Reply:
@jkotas 
I update the previous comment about the test-case name, `src/tests/Interop/StructMarshalling/PInvoke/MarshalStructAsLayoutSeq.cs`

Reply:
I will update this PR later for amending the LA-ABI after merged the implementation of `CEEInfo::getLoongArch64PassStructInRegisterFlags` and `MethodTable::GetLoongArch64PassStructInRegisterFlags` which I had tested on LoongArch64-linux by the SDK6.0.102 with the same test-cases passed.

Reply:
The special casing of arrays should be only done when `IsNativeValueType() == true`. It should be enough to make this test case work.

This path is for `IsNativeValueType() == false`. It should not have any special casing for arrays.

Reply:
@jkotas

Reply:
@jkotas 
Thanks very much.
I will amend this.

Reply:
Yesterday I just only runned all the release test cases as release is faster and all test had the same passed.
Today there are two assert error within the debug, so I update it again after amend to fix them.

Reply:
I had finished the tests on my local machine to test the LoongArch64.

Please give me some advices if you review further.
Thanks